### PR TITLE
CSPL-1625: App framework unit test

### PR DESCRIPTION
**Pull request by vivekr-splunk**
few more unit test case for controller code base

### CODE REVIEW

File: controllers/licensemaster_controller_test.go [hunk 1]
```diff
@@ -39,40 +37,123 @@ var _ = Describe("LicenseMaster Controller", func() {
...
+			}
+			// econcile for the first time err is resource not found
+			_, err := instance.Reconcile(ctx, request)
+			Expect(err).ToNot(HaveOccurred())
+			// create resource first adn then reconcile for the first time
+			ssSpec := testutils.NewLicenseMaster("test", namespace, "image")
+			Expect(c.Create(ctx, ssSpec)).Should(Succeed())
+			// reconcile with updated annotations for pause
+			annotations := make(map[string]string)
+			annotations[enterprisev3.ClusterManagerPausedAnnotation] = ""
```
[the remark this answers is not in the document]
Comment (vivekr-splunk) [reply controllers/licensemaster_controller_test.go:113]: fixed

File: .github/workflows/int-test-workflow.yml [hunk 2]
```diff
@@ -4,7 +4,7 @@ on:
     branches:
       - develop
       - master
-      - feature-appframework
+      - app-framework-unit-test
```
[the remark this answers is not in the document]
Comment (vivekr-splunk) [reply .github/workflows/int-test-workflow.yml:7]: will be removed once intergartion test succeeds

**Review by jryb — APPROVED** [review]

File: pkg/splunk/enterprise/indexercluster.go [hunk 3]
```diff
@@ -103,10 +106,12 @@ func ApplyIndexerCluster(ctx context.Context, client splcommon.ControllerClient,
 	} else {
 		cr.Status.ClusterMasterPhase = splcommon.PhaseError
 	}
-	mgr := indexerClusterPodManager{log: scopedLog, cr: cr, secrets: namespaceScopedSecret, newSplunkClient: splclient.NewSplunkClient}
+
+	//mgr := indexerClusterPodManager{log: scopedLog, cr: cr, secrets: namespaceScopedSecret, newSplunkClient: splclient.NewSplunkClient}
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/indexercluster.go:110]: Remove comment if not needed

File: pkg/splunk/enterprise/indexercluster.go [hunk 4]
```diff
@@ -103,10 +106,12 @@ func ApplyIndexerCluster(ctx context.Context, client splcommon.ControllerClient,
...
+	mgr := NewIndexerClusterPodManager(scopedLog, cr, namespaceScopedSecret, splclient.NewSplunkClient)
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/indexercluster.go:111]: `NewIndexerClusterPodManager` needs to be renamed as indexerClusterPodManager?

File: pkg/splunk/enterprise/indexercluster.go [hunk 5]
```diff
@@ -231,6 +241,16 @@ type indexerClusterPodManager struct {
 	newSplunkClient func(managementURI, username, password string) *splclient.SplunkClient
 }
 
+// NewIndexerClusterPodManager function to create pod manager this is added to write unit test case
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/indexercluster.go:244]: same here

File: pkg/splunk/enterprise/indexercluster.go (hunk 3 above)
Comment (vivekr-splunk) [reply pkg/splunk/enterprise/indexercluster.go:110]: fixed

File: pkg/splunk/enterprise/indexercluster_test.go [hunk 6]
```diff
@@ -1183,3 +1195,504 @@ func TestGetIndexerClusterList(t *testing.T) {
...
+			Name:      "splunk-test-cluster-master-headless",
+			Namespace: "default",
+		},
+	} */
+
+	// simulate service
+	//c.Create(ctx, cservice)
+
+	// simulate create stateful set
+	//c.Create(ctx, cstatefulset)
```
[the remark this answers is not in the document]
Comment (vivekr-splunk) [reply pkg/splunk/enterprise/indexercluster_test.go]: fixed

File: pkg/splunk/enterprise/monitoringconsole_test.go [hunk 7]
```diff
@@ -679,3 +691,491 @@ func TestMonitoringConsoleGetAppsListForAWSS3ClientShouldFail(t *testing.T) {
 		t.Errorf("GetAppsList should return an empty response since we have empty objects in MockAWSS3Client")
 	}
 }
+
+func TestMonitoringConnsoleWithReadyState(t *testing.T) {
+
+	mclient := &spltest.MockHTTPClient{}
+	type Entry1 struct {
+		Content splclient.ClusterMasterInfo `json:"content"`
+	}
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/monitoringconsole_test.go:700]: why do we need clustermaster info here?

File: pkg/splunk/enterprise/monitoringconsole_test.go [hunk 8]
```diff
@@ -679,3 +691,491 @@ func TestMonitoringConsoleGetAppsListForAWSS3ClientShouldFail(t *testing.T) {
...
+
+	response1, err := json.Marshal(apiResponse1)
+	response2, err := json.Marshal(apiResponse2)
+	wantRequest1, _ := http.NewRequest("GET", "https://splunk-test-cluster-master-service.default.svc.cluster.local:8089/services/cluster/master/info?count=0&output_mode=json", nil)
+	wantRequest2, _ := http.NewRequest("GET", "https://splunk-test-cluster-master-service.default.svc.cluster.local:8089/services/cluster/master/peers?count=0&output_mode=json", nil)
+	mclient.AddHandler(wantRequest1, 200, string(response1), nil)
+	mclient.AddHandler(wantRequest2, 200, string(response2), nil)
+
+	// mock the verify RF peer funciton
+	VerifyRFPeers = func(ctx context.Context, mgr indexerClusterPodManager, client splcommon.ControllerClient) error {
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/monitoringconsole_test.go:756]: Same. Can you explain why do we need indexer cluster-info here

File: pkg/splunk/enterprise/monitoringconsole_test.go [hunk 9]
```diff
@@ -679,3 +691,491 @@ func TestMonitoringConsoleGetAppsListForAWSS3ClientShouldFail(t *testing.T) {
...
+	}
+
+	err = c.Get(ctx, namespacedName, clustermaster)
+	if err != nil {
+		t.Errorf("Unexpected get cluster manager %v", err)
+		debug.PrintStack()
+	}
+
+	// call reconciliation
+	_, err = ApplyClusterManager(ctx, c, clustermaster)
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/monitoringconsole_test.go:977]: This all is not necessary. MC is like vanilla standalone

File: pkg/splunk/enterprise/monitoringconsole_test.go [hunk 10]
```diff
@@ -679,3 +691,491 @@ func TestMonitoringConsoleGetAppsListForAWSS3ClientShouldFail(t *testing.T) {
...
+	httpmock.Reset()
+	fixture := `{"status":{"message": "Your message", "code": 200}}`
+	_ = httpmock.NewStringResponder(200, fixture)
+	fakeURL := "https://api.mybiz.com/articles.json"
+
+	// fetch the article into struct
+	client := resty.New()
+	_, err = client.R().Get(fakeURL)
+
+	//monitoringconsole.Status.Initialized = true
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/monitoringconsole_test.go:1171]: Remove comments

File: pkg/splunk/enterprise/searchheadcluster_test.go [hunk 11]
```diff
@@ -1240,3 +1251,780 @@ func TestGetSearchHeadClusterList(t *testing.T) {
...
+	mclient.AddHandler(wantRequest4, 200, string(response2), nil)
+	mclient.AddHandler(wantRequest5, 200, string(response1), nil)
+	mclient.AddHandler(wantRequest6, 200, string(response3), nil)
+
+	mclient.AddHandler(wantRequest7, 200, string(response2), nil)
+	mclient.AddHandler(wantRequest8, 200, string(response1), nil)
+	mclient.AddHandler(wantRequest9, 200, string(response3), nil)
+
+	// mock the verify RF peer funciton
+	VerifyRFPeers = func(ctx context.Context, mgr indexerClusterPodManager, client splcommon.ControllerClient) error {
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/searchheadcluster_test.go:1419]: Not needed for SHC

File: pkg/splunk/enterprise/searchheadcluster_test.go [hunk 12]
```diff
@@ -1240,3 +1251,780 @@ func TestGetSearchHeadClusterList(t *testing.T) {
...
+
+	// appFrameworkSpec: AppSource settings, Poll Interval, volumes, appSources on volumes
+	appFrameworkSpec := enterpriseApi.AppFrameworkSpec{
+		Defaults:             appSourceDefaultSpec,
+		AppsRepoPollInterval: int64(60),
+		VolList:              volumeSpec,
+		AppSources:           appSourceSpec,
+	}
+
+	// create clustermaster custom resource
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/searchheadcluster_test.go:1487]: @vivekr-splunk can you please comment here. Are we testing the SHC ready state that is also connected with the CM?

File: pkg/splunk/enterprise/clustermaster_test.go [hunk 13]
```diff
@@ -996,7 +1003,7 @@ func TestGetClusterManagerList(t *testing.T) {
 	// Invalid scenario since we haven't added clustermanager to the list yet
 	_, err := getClusterManagerList(ctx, client, &cm, listOpts)
 	if err == nil {
-		t.Errorf("getNumOfObjects should have returned error as we haven't added standalone to the list yet")
+		t.Errorf("getNumOfObjects should have returned error as we haven't added clustermaster to the list yet")
```
[the remark this answers is not in the document]
Comment (vivekr-splunk) [reply pkg/splunk/enterprise/clustermaster_test.go]: fixed

File: pkg/splunk/enterprise/monitoringconsole_test.go (hunk 7 above)
Comment (vivekr-splunk) [reply pkg/splunk/enterprise/monitoringconsole_test.go:697]: removed

File: pkg/splunk/enterprise/monitoringconsole_test.go (hunk 9 above)
Comment (vivekr-splunk) [reply pkg/splunk/enterprise/monitoringconsole_test.go]: removed

File: pkg/splunk/enterprise/searchheadcluster_test.go (hunk 11 above)
Comment (vivekr-splunk) [reply pkg/splunk/enterprise/searchheadcluster_test.go:1418]: removed

File: pkg/splunk/enterprise/monitoringconsole_test.go (hunk 8 above)
Comment (kashok-splunk) [reply pkg/splunk/enterprise/monitoringconsole_test.go:753]: We need to remove these :)

File: pkg/splunk/enterprise/monitoringconsole_test.go (hunk 7 above)
Comment (kashok-splunk) [reply pkg/splunk/enterprise/monitoringconsole_test.go:697]: I believe you plan to remove this? correct?

File: pkg/splunk/enterprise/monitoringconsole_test.go [hunk 14]
```diff
@@ -679,3 +688,263 @@ func TestMonitoringConsoleGetAppsListForAWSS3ClientShouldFail(t *testing.T) {
...
+	}
+	err = c.Status().Update(ctx, monitoringconsole)
+	if err != nil {
+		t.Errorf("Unexpected error while running reconciliation for cluster master with app framework  %v", err)
+		debug.PrintStack()
+	}
+
+	err = c.Get(ctx, namespacedName, monitoringconsole)
+	if err != nil {
+		t.Errorf("Unexpected get indexer cluster %v", err)
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/monitoringconsole_test.go:866]: Update message

File: pkg/splunk/enterprise/monitoringconsole_test.go [hunk 15]
```diff
@@ -679,3 +688,263 @@ func TestMonitoringConsoleGetAppsListForAWSS3ClientShouldFail(t *testing.T) {
...
+		debug.PrintStack()
+	}
+
+	// call reconciliation
+	_, err = ApplyMonitoringConsole(ctx, c, monitoringconsole)
+	if err != nil {
+		t.Errorf("Unexpected error while running reconciliation for cluster master with app framework  %v", err)
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/monitoringconsole_test.go:873]: Update message

File: pkg/splunk/enterprise/monitoringconsole_test.go [hunk 16]
```diff
@@ -679,3 +688,263 @@ func TestMonitoringConsoleGetAppsListForAWSS3ClientShouldFail(t *testing.T) {
...
+		debug.PrintStack()
+	}
+
+	stNamespacedName := types.NamespacedName{
+		Name:      "splunk-test-monitoring-console",
+		Namespace: "default",
+	}
+	err = c.Get(ctx, stNamespacedName, statefulset)
+	if err != nil {
+		t.Errorf("Unexpected get indexer cluster %v", err)
```
Comment (kashok-splunk) [inline pkg/splunk/enterprise/monitoringconsole_test.go:926]: Update message

File: pkg/splunk/enterprise/searchheadcluster_test.go (hunk 11 above)
Comment (kashok-splunk) [reply pkg/splunk/enterprise/searchheadcluster_test.go:1418]: Don't need this

**Review by kashok-splunk — CHANGES_REQUESTED** [review]
need to remove some code

File: pkg/splunk/enterprise/configuration.go [hunk 17]
```diff
@@ -1173,9 +1173,17 @@ func ValidateAppFrameworkSpec(ctx context.Context, appFramework *enterpriseApi.A
...
 	// check whether the temporary volume to download apps is mounted or not on the operator pod
-	if _, err := os.Stat(splcommon.AppDownloadVolume); os.IsNotExist(err) {
-		scopedLog.Error(err, "Volume needs to be mounted on operator pod to download apps. Please mount it as a separate volume on operator pod.", "volume path", splcommon.AppDownloadVolume)
+	if os.IsNotExist(err) {
+		scopedLog.Error(err, "Volume needs to be mounted on operator pod to download apps. Please mount it as a separate volume on operator pod.", "volume path", appDownloadVolume)
+		return err
+	}
+
+	if err == nil && !t.IsDir() {
+		scopedLog.Error(err, "Volume needs to be mounted on operator pod to download apps. Please mount it as a separate volume on operator pod.", "volume path", appDownloadVolume)
```
Comment (sgontla) [inline pkg/splunk/enterprise/configuration.go:1186]: As discussed, plz backup and restore, as that helps other test cases to expect a proper value.

File: pkg/splunk/enterprise/configuration_test.go [hunk 18]
```diff
@@ -611,6 +611,13 @@ func TestValidateAppFrameworkSpec(t *testing.T) {
 	var err error
 	ctx := context.TODO()
 
+	// remove the AppDownloadVolume if exist just to make sure
+	// previous test case have not created directory
+	err = os.RemoveAll(splcommon.AppDownloadVolume)
+	if err != nil {
+		t.Errorf("unable to delete directory %s", splcommon.AppDownloadVolume)
+	}
+
```
Comment (sgontla) [inline pkg/splunk/enterprise/configuration_test.go:620]: As discussed, plz backup and restore, as that helps other test cases to expect a proper value.

**Review by sgontla — APPROVED** [review]

File: pkg/splunk/enterprise/monitoringconsole_test.go (hunk 14 above)
Comment (vivekr-splunk) [reply pkg/splunk/enterprise/monitoringconsole_test.go]: fixed

File: pkg/splunk/enterprise/monitoringconsole_test.go (hunk 15 above)
Comment (vivekr-splunk) [reply pkg/splunk/enterprise/monitoringconsole_test.go]: fixed

File: pkg/splunk/enterprise/monitoringconsole_test.go (hunk 16 above)
Comment (vivekr-splunk) [reply pkg/splunk/enterprise/monitoringconsole_test.go]: fixed

**Review by kashok-splunk — APPROVED** [review]

**Review by smohan-splunk — APPROVED** [review]

**Pull request merged**